README.md
## idkit_gshared

#### 1. Briefly

This package is for global data sharing of Flutter projects. The following is a detailed introduction and use of its methods.

#### 2. Use

- Preconditions
  First create a global shared data management object，As follows:
  ```dart
  final IDKitGShared gShared = IDKitGShared();
  ```
- Basic sharing

  ### Register

  ###### 1.Register int、list、map......

  ```dart
  gShared.register<int>(520);
  gShared.register<int>(1314, mark: 'love');
  ```

  ###### 2. Register object

  ```dart
  gShared.register(
      const TestInfo('Asynchronous registration'),
  );
  gShared.register(
      const TestInfo('Asynchronous registration - Mark'),
      mark: 'mark',
  );
  ```

  ###### 3. Asynchronous registration

  ```dart
  gShared.register<Future<String>>(
      Future.delayed(const Duration(seconds: 1), () {
      return '110';
      }),
  );
  gShared.register<Future<TestInfo>>(
      Future.delayed(const Duration(seconds: 1), () {
      return const TestInfo('120');
      }),
      mark: 'mark',
  );
  ```

  ### Read

  ###### 1. Read int、list、map......

  ```dart
  gShared.read<int>() -> int?
  gShared.read<int>(mark: 'love') -> int?
  ```

  ###### 2. Read object

  ```dart
  gShared.read<TestInfo>() -> TestInfo?
  gShared.read<TestInfo>(mark: 'mark') -> TestInfo?
  ```

  ###### 3. Read data asynchronously

  ```dart
  FutureBuilder<String>(
      future: gShared.read<Future<String>>(),
      initialData: '000',
      builder: (_, AsyncSnapshot<String> snapshot) {
      return Text('${snapshot.data}');
      },
  ),
  FutureBuilder<String>(
      future: gShared.read<Future<String>>(mark: 'mark'),
      initialData: '000 - mark',
      builder: (_, AsyncSnapshot<String> snapshot) {
      return Text('${snapshot.data}');
      },
  ),
  ```

  ### Update data

  ```dart
  gShared.update<String>((value) => 'I am a baby!');
  gShared.update<String>((value) => 'I am a baby!',mark:'mark');
  ```

  ### unRegister

  ```dart
  gShared.unRegister<String>();
  gShared.unRegister<String>(mark: 'mark');
  ```

  or

  ```dart
  gShared.unRegisterAll();
  gShared.unRegisterAll(listen: true);
  ```

- Listen sharing

  ### Register

  ```dart
  gShared.registerListen<String>();
  gShared.registerListen<String>(mark: 'mark');
  ```

  ### Watch

  ```dart
  StreamBuilder<String?>(
     stream: gShared.watch<String>(),
     initialData: 'listen - 0',
     builder: (_, s) {
        final String? a = s.data;
        return Text(a ?? 'listen - 0');
     },
  ),
  StreamBuilder<String?>(
    stream: gShared.watch<String>(mark: 'mark'),
    initialData: 'listen - 0',
    builder: (_, s) {
        final String? a = s.data;
        return Text(a ?? 'listen - 0');
    },
  )
  ```

  ### unRegister listen

  ```dart
  gShared.unRegisterListen<String>();
  gShared.unRegisterListen<String>(mark: 'mark');
  ```

  ### Convert

  ```dart
  gShared.convertListen<int>();
  gShared.convertListen<int>(mark:'mark');
  ```

### 3. Summarize

This version of the package covers some of the above methods. If you don't understand anything or don't understand how to use it, you can leave me a message!


CHANGELOG.md
## 0.0.1

##### 1. Describe

This package is a solution to the relative troublesomeness of global data sharing in Flutter projects and requires a lot of code to be written. This package is easy to use, and the main process includes 3 main steps such as registration, reading, and logout. This package also supports asynchronous registration and monitoring of data.

##### 2. Extension methods and descriptions

- The factory method generates a globally shared management class.

  ```dart
  IDKitGShared factory IDKitGShared() => ...
  ```

- Register and manage the data objects to be shared.
  ```dart
  void register<T extends Object>( T? instance,{
      String? mark,
  })
  ```
- Get shared data object.

  ```dart
  T? read<T extends Object>({
      String? mark,
      bool disposable = false,
  })
  ```

- Unregister managed shared data objects.

  ```dart
  void unRegister<T extends Object>({
      String? mark,
  })
  ```

- Unregister all managed shared data objects.

  ```dart
  void unRegisterAll({
      bool listen = false,
  })
  ```

- Update shared data objects.

  ```dart
  void update<T extends Object>(
      T? Function(T? value) update, {
      String? mark,
  })
  ```

- Conveniently convert non-listening shared data objects into listening.

  ```dart
  void convertListen<T extends Object>({
      String? mark,
      bool broadcast = false,
      bool sync = false,
  })
  ```

- Registering a shared data object for listening mode.

  ```dart
  void registerListen<T extends Object>({
      String? mark,
      bool broadcast = false,
      bool sync = false,
  })
  ```

- Listen to shared visible listen data objects.

  ```dart
  Stream<T?>? watch<T extends Object>({
      String? mark,
  })
  ```

- Send listening shared data.

  ```dart
  void add<T extends Object>(
      T value, {
      String? mark,
  })
  ```

- Log out subscriber.

  ```dart
  void unRegisterListen<T extends Object>({
      String? mark,
  })
  ```

##### 3. Instructions for use

If you want to know more about this package, please check the [README](https://github.com/zhoushuangjian001/idkit_gshared/blob/master/README.md).


lib/src/gshared_main.dart
import 'dart:async';
import 'package:flutter/material.dart';
import 'package:idkit_gshared/src/gshared_func.dart';

@immutable
class IDKitGShared {
  // Record shared data objects.
  late final Map<String, dynamic> _recordPool;
  // Record shared listening data object.
  late final Map<String, StreamController> _recordListenPool;
  // Shared management object construction.
  factory IDKitGShared() => _instance();
  static IDKitGShared _instance() => IDKitGShared._init();
  IDKitGShared._init() {
    _recordPool = {};
    _recordListenPool = {};
  }

  /// Register shared objects.
  ///
  /// [instance] --
  /// Shared object to register.
  ///
  /// [mark] --
  /// The unique identifier of the shared object to be registered.
  /// If there is only one shared object of the same type in the project,
  /// it can be used to distinguish it. Otherwise, it needs to be used to distinguish it.
  void register<T extends Object>(
    T? instance, {
    String? mark,
  }) {
    final Iterable<String> keys = _recordPool.keys;
    final String key = keyGet<T>(mark);
    if (keys.contains(key)) {
      _recordPool.update(key, (value) => instance);
    } else {
      _recordPool.putIfAbsent(key, () => instance);
    }
  }

  /// Get global shared object.
  ///
  /// [mark] --
  /// The unique identifier of the shared object to be registered.
  /// If there is only one shared object of the same type in the project,
  /// it can be used to distinguish it. Otherwise, it needs to be used to distinguish it.
  ///
  /// [disposable] --
  /// Gets whether the global shared object is one-time consumption.
  /// Even if you use it once, you can't use it again next time.
  T? read<T extends Object>({
    String? mark,
    bool disposable = false,
  }) {
    final String key = keyGet<T>(mark);
    if (disposable) {
      _recordPool.remove(key);
    }
    return instanceGet(_recordPool, key);
  }

  /// Unregister the specified global shared object.
  ///
  /// [mark] --
  /// The unique identifier of the shared object to be registered.
  /// If there is only one shared object of the same type in the project,
  /// it can be used to distinguish it. Otherwise, it needs to be used to distinguish it.
  void unRegister<T extends Object>({
    String? mark,
  }) {
    final Iterable<String> keys = _recordPool.keys;
    late String key = keyGet<T>(mark);
    if (keys.contains(key)) {
      _recordPool.remove(key);
    }
  }

  /// Unregister all global shared objects.
  ///
  /// [listen] --
  /// Distinguish between listening and non-listening global shared objects
  void unRegisterAll({
    bool listen = false,
  }) {
    _recordPool.clear();
    if (listen) {
      _recordListenPool.clear();
    }
  }

  /// Updates the specified global shared object.
  ///
  /// [update] --
  /// Callback event for updating shared data.
  ///
  /// [mark] --
  /// The unique identifier of the shared object to be registered.
  /// If there is only one shared object of the same type in the project,
  /// it can be used to distinguish it. Otherwise, it needs to be used to distinguish it.
  void update<T extends Object>(
    T? Function(T? value) update, {
    String? mark,
  }) {
    final String key = keyGet<T>(mark);
    final T? _instance = instanceGet(_recordPool, key);
    final T? item = update.call(_instance);
    if (_instance == null) {
      _recordPool.putIfAbsent(key, () => item);
    } else {
      _recordPool.update(key, (value) => item);
    }
  }

  /// Convert an object that has no listeners to a listener object.
  ///
  /// [mark] --
  /// Is the unique identifier of the shared data object before the query.
  ///
  /// [broadcast] --
  /// Listener broadcast form of shared object.
  ///
  /// [sync] --
  /// Whether the broadcast event should be delayed.
  void convertListen<T extends Object>({
    String? mark,
    bool broadcast = false,
    bool sync = false,
  }) {
    final String key = keyGet<T>(mark);
    final bool isExist = _recordPool.keys.contains(key);
    final StreamController<T> streamController = broadcast ? StreamController<T>.broadcast(sync: sync) : StreamController<T>(sync: sync);
    _recordListenPool[key] = streamController;
    if (isExist) {
      _recordPool.remove(key);
    }
  }

  /// Register a global shared listener object.
  ///
  /// [mark] --
  /// Is the unique identifier of the shared data object before the query.
  ///
  /// [broadcast] --
  /// Listener broadcast form of shared object.
  ///
  /// [sync] --
  /// Whether the broadcast event should be delayed.
  void registerListen<T extends Object>({
    String? mark,
    bool broadcast = false,
    bool sync = false,
  }) {
    final String key = keyGet<T>(mark);
    final StreamController<T> streamController = broadcast ? StreamController<T>.broadcast(sync: sync) : StreamController<T>(sync: sync);
    _recordListenPool[key] = streamController;
  }

  /// Subscribing to a shared listener data object.
  ///
  /// [mark] --
  /// Is the unique identifier of the shared data object before the query.
  Stream<T>? watch<T extends Object>({
    String? mark,
  }) {
    final String key = keyGet<T>(mark);
    final Iterable keys = _recordListenPool.keys;
    if (keys.contains(key)) {
      final StreamController<T>? gstreamController = instanceListenGet<T>(_recordListenPool, key);
      return gstreamController?.stream;
    }
    return null;
  }

  /// Send shared listening data.
  ///
  /// [value] --
  /// Shared listening data to send.
  ///
  /// [mark] --
  /// Check the unique identifier of the new sending carrier.
  void add<T extends Object>(
    T value, {
    String? mark,
  }) {
    final String key = keyGet<T>(mark);
    final Iterable keys = _recordListenPool.keys;
    if (keys.contains(key)) {
      final StreamController<T>? gstreamController = instanceListenGet(_recordListenPool, key);
      gstreamController?.add(value);
    }
  }

  /// Remove the specified listener object.
  ///
  /// [mark] --
  /// Query the unique identifier of the monitored object.
  void unRegisterListen<T extends Object>({
    String? mark,
  }) {
    final String key = keyGet<T>(mark);
    final Iterable keys = _recordListenPool.keys;
    if (keys.contains(key)) {
      final StreamController<T>? gstreamController = instanceListenGet(_recordListenPool, key);
      gstreamController?.close();
      _recordListenPool.remove(key);
    }
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:idkit_gshared/idkit_gshared.dart';

/// Global Shared Simple Interest
final IDKitGShared gShared = IDKitGShared();

/// Test class
class TestInfo {
  final String name;
  const TestInfo(this.name);
}

void main() {
  gShared.register<int>(520);
  gShared.register<int>(1314, mark: 'love');
  gShared.register(
    const TestInfo('Asynchronous registration'),
  );
  gShared.register(
    const TestInfo('Asynchronous registration - Mark'),
    mark: 'r_mark',
  );

  gShared.register<Future<String>>(
    Future.delayed(const Duration(seconds: 1), () {
      return '110';
    }),
  );
  gShared.register<Future<TestInfo>>(
    Future.delayed(const Duration(seconds: 1), () {
      return const TestInfo('120');
    }),
    mark: 'mark',
  );
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('idkit_gshared test'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: Column(
          children: [
            Text('${gShared.read<int>() ?? 0}'),
            Text('${gShared.read<int>(mark: 'love') ?? 0}'),
            Text('${gShared.read<TestInfo>()?.name ?? 0}'),
            Text('${gShared.read<TestInfo>()?.name ?? 0}'),
            FutureBuilder<String>(
              future: gShared.read<Future<String>>(),
              initialData: '000',
              builder: (_, AsyncSnapshot<String> snapshot) {
                return Text('${snapshot.data}');
              },
            ),
            FutureBuilder<String>(
              future: gShared.read<Future<String>>(mark: 'mark'),
              initialData: '000 - mark',
              builder: (_, AsyncSnapshot<String> snapshot) {
                return Text('${snapshot.data}');
              },
            ),
            ElevatedButton(
              child: const Text('Next Step'),
              onPressed: () {
                Navigator.of(context).push(MaterialPageRoute(builder: (_) {
                  return const SecondPage();
                }));
              },
            )
          ],
        ),
      ),
    );
  }
}

class SecondPage extends StatefulWidget {
  const SecondPage({Key? key}) : super(key: key);

  @override
  State<SecondPage> createState() => _SecondPageState();
}

class _SecondPageState extends State<SecondPage> {
  late int count = 0;

  @override
  Widget build(BuildContext context) {
    gShared.registerListen<String>();
    gShared.registerListen<String>(mark: 'mark');
    return Scaffold(
      appBar: AppBar(
        title: const Text('idkit_gshared:Multi-layer data sharing'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: Column(
          children: <Widget>[
            const SizedBox(
              height: 20,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: [
                Text(gShared.read<String>(mark: 'wz') ?? 'defaulu value: network'),
                StreamBuilder<String?>(
                  stream: gShared.watch<String>(),
                  initialData: 'listen - 0',
                  builder: (_, s) {
                    final String? a = s.data;
                    return Text(a ?? 'listen - 0');
                  },
                ),
                StreamBuilder<String?>(
                  stream: gShared.watch<String>(mark: 'mark'),
                  initialData: 'listen - 0',
                  builder: (_, s) {
                    final String? a = s.data;
                    return Text(a ?? 'listen - 0');
                  },
                )
              ],
            ),
            const SizedBox(
              height: 20,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceEvenly,
              children: <Widget>[
                ElevatedButton(
                  onPressed: () {
                    count++;
                    gShared.add<String>('listen - $count', mark: 'listen');
                  },
                  child: const Text('Add Count + 1'),
                ),
                ElevatedButton(
                  onPressed: () {
                    Navigator.of(context).push(MaterialPageRoute(builder: (_) {
                      return const ThreePage();
                    }));
                  },
                  child: const Text('Next Step Button'),
                ),
                ElevatedButton(
                  onPressed: () {
                    gShared.unRegister<String>();
                    gShared.unRegister<String>(mark: 'mark');
                    gShared.unRegisterAll();
                    gShared.unRegisterAll(listen: true);
                    gShared.update<String>((value) => 'I am a baby!');
                  },
                  child: const Text('Update Data'),
                )
              ],
            )
          ],
        ),
      ),
    );
  }
}

class ThreePage extends StatefulWidget {
  const ThreePage({Key? key}) : super(key: key);

  @override
  State<ThreePage> createState() => _ThreePageState();
}

class _ThreePageState extends State<ThreePage> {
  late int count = 0;

  @override
  Widget build(BuildContext context) {
    gShared.convertListen<int>();
    gShared.unRegisterListen<String>();
    gShared.unRegisterListen<String>(mark: 'mark');

    return Scaffold(
      appBar: AppBar(
        title: const Text('Three Page 3'),
      ),
      body: Container(
        alignment: Alignment.center,
        child: Column(
          children: <Widget>[
            Text(gShared.read<String>() ?? ''),
            StreamBuilder<int?>(
              stream: gShared.watch<int>(),
              initialData: 0000,
              builder: (_, s) {
                return Text('${s.data ?? 1111}');
              },
            )
          ],
        ),
      ),
      floatingActionButton: ElevatedButton(
        child: const Text('Add Change + 1'),
        onPressed: () {
          count++;
          gShared.add<int>(count);
        },
      ),
    );
  }
}
